Lib/pathlib.py

```diff
@@ -194,25 +194,24 @@ def resolve(self, path, strict=False):
         s = str(path)
         if not s:
             return os.getcwd()
+        if _getfinalpathname is None:
+            return None  # Means fallback on absolute
+        if strict:
+            return self._ext_to_normal(_getfinalpathname(s))
+        s = path = os.path.abspath(s)
```

Comment:
Hmm, I think we want to call `nt._getfullpathname` instead. `os.path.abspath` will blindly simplify `..` segments in a way that changes the actual path if there are symlinks.

Reply:
They've already said non-strict 🤷‍♂️

Besides, `_getfinalpathname` will fail if it doesn't exist. You want `ntpath.realpath` to get the full non-strict resolution behaviour.

Reply:
I suggested `_getfullpathname`, not `_getfinalpathname`.
And I'm not sure what "non-strict" is supposed to mean here, but IMHO it shouldn't mean "silently return the wrong path".
(yes, symlinks are rare on Windows, etc. Still, we should strive for correct behaviour)

Reply:
You're right, my bad.

>  IMHO it shouldn't mean "silently return the wrong path".

Yeah, I think that's fair. Though I take non-strict to mean "best effort given we can't resolve the path".

One real scenario I hit here is mount points, where I have my `D:` drive also mounted at `C:\Files`. With CWD at `C:\Files\Spam` and a relative path of `..\..\Windows`, the normalised version will be correct (`C:\Windows`) while the left-to-right resolution of `C:\Files\Spam\..\..\Windows` would be incorrect.

So ultimately, it's less a case of "what is the correct behaviour" and more "what is the most predictable behaviour and do we have the tools for users to work around it"?

Reply:
I definitely specified (and implemented, to the best of my abilities ;-)) `resolve` to do the right thing. Similarly, the reason I refused to document `absolute` is that it does the wrong thing, which could end up hurtful.

Reply:
At the outset of `resolve()`, the correct behavior is to call `_getfullpathname`. This handles relative paths and removes ".." components. It also eagerly resolves cases such as "C:/Temp/nul.txt" -> "\\\\.\\nul" that `_getfinalpathname` fails on. From user mode (e.g. `CreateFileW` and `CreateDirectorW`), Windows does not care about symlinks when normalizing a path before opening it. Historically it didn't have them. And most filesystems (e.g. NTFS) in the kernel have no special handling for a component named ".." anyway, except to reject it as an invalid name. So opening "C:\\spam\\symlink\\..\\eggs" is always opened as "C:\\spam\\eggs". 

In the kernel it's only the I/O manager's implementation of symlink reparsing that evaluates ".." components when resolving the target path of a relative symlink. That had to be implemented because they're supposed to work like Unix symlinks (except they dropped the ball on handling "/" as a path separator in relative symlinks). 

> One real scenario I hit here is mount points, where I have my `D:` drive also mounted at `C:\Files`. With CWD at `C:\Files\Spam` and a relative path of `..\..\Windows`, the normalised version will be correct (`C:\Windows`) while the left-to-right resolution of `C:\Files\Spam\..\..\Windows` would be incorrect.

The left to right resolution when done properly would be correct. For example, consider what the kernel does for relative symlinks. Say you have a mountpoint "C:\\Files" and a relative symlink "C:\\Files\\Spam\\Symlink" that targets "..\\..\\Windows". Path parsing in the kernel, which is left to right, takes into account that "Files" is a mountpoint (`IO_REPARSE_TAG_MOUNT_POINT`) and not a symlink (`IO_REPARSE_TAG_SYMLINK`). A mountpoint remains in the resolved path while parsing a path, just like a regular directory. The I/O manager follows "Files" to the target volume path, but "\\Device\\HarddiskVolume2\\Files" is retained in the resolved path (given "C:" -> "\\Device\\HarddiskVolume2").

When the filesystem reaches the "Symlink" reparse point and returns `STATUS_REPARSE` with the contents of the reparse point, the I/O manager reparses the relative symlink target against the resolved path "\\Device\\HarddiskVolume2\\Files\\Spam\\Symlink", and this yields the correct target path "\\Device\\HarddiskVolume2\\Windows".

OTOH, if "C:\\Files" is a symlink to "D:\\" or even a symlink to the "\\\\?\\Volume{GUID}\\" mountpoint (it's the type of reparse point that matters, not its target), then the previously resolved path is no longer remembered. In this case, the filesystem still reaches "\Spam\Symlink" and returns `STATUS_REPARSE` with the reparse data, just as before, but this time the I/O manager fails the open with `ERROR_INVALID_REPARSE_DATA`. Let's say the resolved path is "\\Device\\HarddiskVolume10\\Spam\\Symlink" (given "D:" -> "\\Device\\HarddiskVolume10"). In this case there's one too many ".." components, which Windows does not allow. (If allowed literally, it would traverse into the object namespace.) 

So "Symlink" is valid when "C:\\Files" is a mountpoint, and invalid when "C:\\Files" is a symlink. You can replicate similar behavior for bind mountpoints vs symlinks in Linux, except Linux terminates excess ".." components at root.

---

If you're manually parsing a path in order to aggressively follow broken symlinks as far as possible, then you can cheat instead of replicating all the hard work the kernel has to do if what you want is a "final" path as the real path. Traverse the path in reverse, and call `CreateFileW` with `FILE_FLAG_OPEN_REPARSE_POINT` until it succeeds. Call `GetFinalPathNameByHandleW` on that file handle and check if it's a reparse point via `GetFileInformationByHandleEx`: `FileAttributeTagInfo`. 

If final component that's accessible is a mountpoint or symlink, and the final path is not a UNC path (i.e. it's local) or the file is a symlink with a relative or remote target path, then it is safe to read and apply the reparse target. It is not safe to do this for a remote UNC path if the target is a local drive path. In particular it would be dangerous nonsense  to do this for a remote mountpoint (e.g. targeting some remote drive "Q:", which could be nothing or anything in the local context), which is mandated to only be evaluated on the server side. The same applies to R2L (remote to local) symlinks, which any sane system symlink policy will disable, so manual evaluation should just assume that R2L symlinks are disabled (i.e. only allow L2L, L2R, and R2R).